### PR TITLE
Fix inactive Calibrate button (#105): never silently default a request parameter

### LINKED ISSUE #105: inactive Calibrate button
The Calibrate button is inactive when I'm trying to calibrate the sample data in issue #101:

Image

### PR DESCRIPTION
**Pull request by cliu238**
Closes #105. Unblocks investigation of #101.

## What was wrong

The Calibrate button was permanently disabled for child uploads, because the cause-mapping preview was scoring them against the **neonate** cause list.

`previewMapping()` sent `age_group` as a multipart form field. plumber gives a text part no Content-Type and no filename, so `parser_picker()` falls through to `parser_text(parseQS)`; `parseQS` reads the raw part value as a query string, and a bare `child` contains no `=`, so it parses to an **empty list**. `POST /jobs/preview` could not distinguish that from an absent parameter and substituted `"neonate"` (`plumber.R:392`).

`hiv`, `other_infections` and `severe_malnutrition` are not neonate causes, so all three reported as unrecognized, `has_errors` became true, and `JobForm.jsx:468` disabled submission. `submitJob` was never affected, because it has always sent its scalars via `URLSearchParams` — which is why job submission worked while the preview did not.

## How it was confirmed

Against the **live deployed backend** (pre-fix), with the CSV attached to #101 — same file, same server, only the transport varied:

| Probe | Result |
|---|---|
| `-F age_group=child` (multipart) | 1334/2383; unrecognized `hiv`=162, `other_infections`=720, `severe_malnutrition`=167 — matches the #105 screenshot item for item |
| `?age_group=child` (query string) | 2383/2383, **zero** unrecognized, all 7 causes identity-mapped |
| `age_group` omitted entirely | **byte-identical to the multipart probe** |

That last identity is the decisive part: sending the value as multipart is indistinguishable from not sending it, so the value is *destroyed*, not mis-parsed.

Reproduced locally too — `preview_cause_mapping()` on the real 2383-record CSV as `neonate` yields exactly 1334/2383 with `diarrhea`→`sepsis_meningitis_inf` (473) and `malaria`→`sepsis_meningitis_inf` (201); `sepsis_meningitis_inf` is not even a child broad cause.

## Changes

- **`previewMapping()` sends `age_group` in the query string**, matching `submitJob`. This one change fixes #105 against the currently deployed backend, so **no backend redeploy is needed to unblock users**.
- **New parameter helpers in `jobs/utils.R`.** A parameter that determines *what science ran* is required and never guessed; a tuning knob keeps its documented default but rejects invalid values rather than letting `as.integer("abc")` or `as.logical("yes")` silently become `NA`.
- **`POST /jobs` hardened.** The same silent `age_group → "neonate"` fallback was sitting on the endpoint that *creates* the calibration, masked only because `client.js` happens to use `URLSearchParams` there. A lost value there yields a silently wrong calibration rather than a visibly disabled button.

  | Treatment | Parameters | Why |
  |---|---|---|
  | Required, never guessed | `job_type`, `algorithm`, `age_group`, `country` | Determine what science ran. `country` selects the CHAMPS misclassification matrix stratum, so a lost value calibrates against the wrong country |
  | Default kept, value validated | `calib_model_type`, `ensemble`, `n_mcmc`, `n_burn`, `n_thin` | Genuine tuning knobs with defensible conventional defaults |

- **`JobForm` surfaces a top-level preview error** instead of swallowing it into an empty report set that would look like a clean preview.

### Two further defects found in the same code

- **`POST /jobs` error handling never worked.** `return()` inside a `tryCatch` error handler exits only the handler, so a parse failure fell through and the endpoint kept running with unresolved parameters.
- **A repeated parameter with conflicting values silently took the first one.** That is another silent guess; now rejected.

### Docs that taught the bug

`backend/README.md` showed `-F "job_type=..." -F "algorithm=..." -F "age_group=..." -F "country=..."` — precisely the pattern plumber discards. Corrected to query-string transport with a required-vs-optional split, along with three stale `-F "age_group=..."` curl examples in the cause-mapping-preview plan.

## Verification

| Suite | Result |
|---|---|
| New backend parameter validation (section 2g) | 38/38 |
| Live endpoint resolution block | 20/20 |
| Issue #105 mapping assertions, real CSV | 17/17 |
| Frontend, as CI runs it | 240/240 across 32 files |

The endpoint suite **extracts the `resolved <- tryCatch(...)` block from `plumber.R` at runtime** and executes it against synthetic requests, so it cannot drift from the shipped code. It confirms the exact payload `submitJob` sends still works (single algorithm and ensemble), each required parameter now errors loudly when lost, and the five knobs still default correctly.

There is also a drift guard asserting `CALIBRATION_COUNTRIES` still equals `names(Mmat_champs$neonate$eava$postmean)`, so a future package change that alters the country strata fails a test instead of silently narrowing validation.

## Known gaps, not introduced here

- **The new backend assertions cannot run in CI.** `tests/test_vacalibration_backend.R` halts earlier at section 2b with `object 'comsamoz_CCVAoutput' not found` — the installed `vacalibration` exposes only `Mmat_champs`, `comsamoz_public_broad` and `comsamoz_public_openVAout`, and the stale name is used at 8 sites with a nested `$neonate$eava` shape, so it needs a dataset migration rather than a rename. Separately, `.github/workflows/test.yml` has no R job. The frontend guard, which covers the actual transport defect, *does* run in CI.
- **No browser confirmation yet.** `JobForm` is behind `ProtectedRoute`, and a local backend cannot be started here (`DBI`/`RPostgres` absent). Worth a click-through after deploy: Children (1–59 months) + EAVA + `sample_eava_1to59m.csv` should show 2383/2383 with Calibrate enabled.
- **`POST /jobs/demo` untouched.** Its plumber-signature defaults are a legitimate pattern for a canned demo, but it still coerces with bare `as.logical()`/`as.integer()`, so garbage can still become `NA` there. Follow-up.

## Scope caveat

**This does not explain #101.** Job submission always delivered `age_group` correctly, so "same results for uncalibrated and calibrated" is a separate defect. This PR only removes the blocker that prevented #101 from being reproduced on the dashboard.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed preview requests so age-group selections are transmitted reliably.
  * Prevented invalid or missing job parameters from being silently defaulted.
  * Improved preview error handling with clear service-unavailable feedback.
  * Added validation for algorithms, countries, calibration settings, and tuning options.

* **Documentation**
  * Clarified parameter requirements and request formatting for job submission and preview.
  * Added updated examples for uploads, previews, and ensemble requests.

### CODE REVIEW

File: backend/jobs/utils.R [hunk 1]
```diff
@@ -387,6 +387,137 @@ extract_top_cod <- function(result) {
...
+optional_count <- function(value, name, default) {
+  v <- param_scalar(value, name)
+  if (!nzchar(v)) return(default)
+  n <- suppressWarnings(as.numeric(v))
+  if (is.na(n) || n < 1 || n != trunc(n)) {
+    stop(sprintf("Invalid '%s' value '%s'. Must be a positive whole number.", name, v),
+         call. = FALSE)
+  }
+  as.integer(n)
+}
```
[the remark this answers is not in the document]
Comment (cliu238) [reply backend/jobs/utils.R:475]: Fixed in e9a9112. `optional_count` now rejects values past `.Machine$integer.max` before the `as.integer()` conversion, so "3e9" errors loudly instead of overflowing to `NA`. Added a regression test alongside the other `optional_count` cases (`tests/test_vacalibration_backend.R`), and verified in a live R session that 3e9 is rejected while the boundary value 2147483647 is still accepted.

File: backend/README.md [hunk 2]
```diff
@@ -31,26 +31,46 @@ curl http://localhost:8000/health
...
-  -F "algorithm=InterVA" \
-  -F "age_group=neonate" \
-  -F "country=Mozambique"
+# With custom data -- scalars in the query string, file as multipart
+curl -X POST "http://localhost:8000/jobs?job_type=pipeline&algorithm=InterVA&age_group=neonate&country=Mozambique" \
+  -F "file=@data.csv"
+
+# Ensemble: algorithm accepts a JSON array
+curl -X POST "http://localhost:8000/jobs?job_type=pipeline&algorithm=%5B%22InterVA%22,%22EAVA%22%5D&age_group=child&country=Kenya&ensemble=true" \
+  -F "file_interva=@interva.csv" -F "file_eava=@eava.csv"
```
[the remark this answers is not in the document]
Comment (cliu238) [reply backend/README.md]: Fixed in e9a9112. Confirmed against `plumber.R`: the per-algorithm file branch fires only for `job_type == "vacalibration"` with 2+ algorithms, so the pipeline example would indeed have created a job with no input file. The example now uses `job_type=vacalibration` (which is also what the dashboard actually submits), with a comment noting that a pipeline ensemble takes a single `-F "file=@data.csv"` instead.

File: docs/superpowers/plans/2026-07-22-cause-mapping-preview.md [hunk 3]
```diff
@@ -188,7 +188,7 @@ git commit -m "feat: preview_cause_mapping report helper for pre-submit cause pr
 
 **Interfaces:**
 - Consumes: `preview_cause_mapping` (Task 1), `save_uploaded_file` (existing, `backend/plumber.R:19`).
-- Produces: `POST /jobs/preview` → JSON `{ reports: { <algo>: <report> } }`, or `{ error: <msg> }`. Accepts the same multipart args as `POST /jobs`: `age_group`, and `file` (single) or `file_interva`/`file_insilicova`/`file_eava` (multi).
+- Produces: `POST /jobs/preview` → JSON `{ reports: { <algo>: <report> } }`, or `{ error: <msg> }`. Files are multipart, exactly as `POST /jobs` takes them: `file` (single) or `file_interva`/`file_insilicova`/`file_eava` (multi). **CORRECTION (issue #105): `age_group` must go in the QUERY STRING, not multipart.** As written here it was sent as a multipart field, which plumber discards (a text part gets no Content-Type, so `parseQS` turns a bare `neonate` into an empty list) — the endpoint then defaulted to `neonate` and mis-scored every child upload.
```
[the remark this answers is not in the document]
Comment (cliu238) [reply docs/superpowers/plans/2026-07-22-cause-mapping-preview.md:191]: Fixed in e9a9112. The Step 1 snippet now carries an explicit **Superseded** callout stating that the silent `age_group <- "neonate"` fallback is the #105 defect and that the shipped code requires `age_group` from the query string via `resolve_age_group()`. Kept the original snippet body intact since the plan is a historical record — the callout says "do not copy this snippet".

File: frontend/src/components/JobForm.jsx [hunk 4]
```diff
@@ -107,7 +107,16 @@ export default function JobForm({ onJobSubmitted }) {
...
+      // A top-level `error` (e.g. a rejected age_group) is not a per-file cause
+      // report, so it must not be swallowed into an empty report set that looks
+      // like "preview clean". Surface it via the same notice as a request
+      // failure. Issue #105.
+      .then(res => {
+        if (cancelled) return;
+        setPreviewPending(false);
+        if (res.error) { setPreviewReports({}); setPreviewUnavailable(true); return; }
+        setPreviewReports(res.reports || {});
+      })
```
[the remark this answers is not in the document]
Comment (cliu238) [reply frontend/src/components/JobForm.jsx:123]: Fixed in e9a9112. `previewUnavailable` (boolean) is now `previewNotice` (string): a top-level `res.error` renders the backend's own message ("Couldn't preview cause mapping: <backend message> Your file will still be validated when you submit."), while a network/server failure in `.catch` keeps the generic service-unavailable wording. Submit stays live in both cases since `/jobs` re-validates. The behavior test now asserts the backend message is shown and the outage wording is absent.

File: frontend/src/components/JobForm.jsx [hunk 5]
```diff
@@ -107,7 +107,16 @@ export default function JobForm({ onJobSubmitted }) {
...
-      .then(res => { if (!cancelled) { setPreviewReports(res.reports || {}); setPreviewPending(false); } })
+      // A top-level `error` (e.g. a rejected age_group) is not a per-file cause
+      // report, so it must not be swallowed into an empty report set that looks
+      // like "preview clean". Surface it via the same notice as a request
+      // failure. Issue #105.
+      .then(res => {
+        if (cancelled) return;
+        setPreviewPending(false);
+        if (res.error) { setPreviewReports({}); setPreviewUnavailable(true); return; }
+        setPreviewReports(res.reports || {});
```
[the remark this answers is not in the document]
Comment (cliu238) [reply frontend/src/components/JobForm.jsx:122]: Fixed in e9a9112 (same finding as CodeRabbit's on this hunk). The component now keeps a `previewNotice` string: backend rejections render the backend's specific error text — including the transport hint — and only genuine request failures (`.catch`) get the hardcoded service-unavailable wording. Behavior test updated to pin this down.

**Comment by cliu238** [comment]
**Re: CodeRabbit's nitpick (plumber.R:236 — return HTTP 400 for parameter-validation failures):** intentionally not changed in this PR. The entire API surface in `plumber.R` uses the 200 + `{error: ...}` payload contract — every error return in `POST /jobs` (ensemble validation, missing files, save failures) and `POST /jobs/preview` behaves this way, and the frontend is built around it: `fetchJson` throws on non-2xx, so a 400 would route a validation rejection into the generic `.catch` outage path instead of the `res.error` branch that (as of e9a9112) surfaces the backend's specific message. Switching one branch to 400 piecemeal would make parameter errors the only class that behaves differently. A wholesale status-code migration (all error returns + `fetchJson` reworked to parse error bodies from non-2xx) is a reasonable follow-up, but out of scope for this fix.

All other review feedback addressed in e9a9112:
- `optional_count` bounds against `.Machine$integer.max` (+ regression test)
- README ensemble example corrected to `job_type=vacalibration`
- Plan doc's Step 1 snippet marked superseded
- `JobForm` renders the backend's preview rejection message instead of mislabeling it as a service outage

Frontend suite as CI runs it: 240/240 across 32 files.
